Use Devices instead of DeviceIndexes in Future

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57354 Inline hooks in ivalue::Future
* **#57353 Use Devices instead of DeviceIndexes in Future**
* #57294 Use Devices instead of DeviceIndexes in TensorPipe agent
* #57345 Rename DeviceGuardImplInteface's getStreamFromPool method
* #57293 Add device() method to c10::Event
* #57292 Add pybind type caster for c10::Device

Even though we merged CUDAFuture into ivalue::Future, the resulting methods still had basically two distinct codepaths (i.e., an "early exit" if `impl_ == nullptr` for CPU, and then some code for CUDA). This works but it risks creating divergence and inconsistencies when the same class is used in those two modes. Ideally we should have the same codepath, and have the stream operations be no-ops for CPU. Luckily, this is exactly what happens when using a CPU DeviceGuardImplInterface!

Hence here I do that, and for convenience I also use c10::Devices instead of c10::DeviceIndexes (like we did in https://github.com/pytorch/pytorch/pull/57294 for RPC).

Differential Revision: [D28100525](https://our.internmc.facebook.com/intern/diff/D28100525/)

lw committed:
Use Devices instead of DeviceIndexes in Future

Even though we merged CUDAFuture into ivalue::Future, the resulting methods still had basically two distinct codepaths (i.e., an "early exit" if `impl_ == nullptr` for CPU, and then some code for CUDA). This works but it risks creating divergence and inconsistencies when the same class is used in those two modes. Ideally we should have the same codepath, and have the stream operations be no-ops for CPU. Luckily, this is exactly what happens when using a CPU DeviceGuardImplInterface!

Hence here I do that, and for convenience I also use c10::Devices instead of c10::DeviceIndexes (like we did in https://github.com/pytorch/pytorch/pull/57294 for RPC).

Differential Revision: [D28100525](https://our.internmc.facebook.com/intern/diff/D28100525/)

[ghstack-poisoned]

lw commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -797,16 +799,14 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   // An upcast pointer to a virtual class which allows us to manipulate events,
   // streams, ... in a generic way, without an explicit dependency on CUDA.
-  const c10::impl::DeviceGuardImplInterface* const impl_ = nullptr;
+  const c10::impl::VirtualGuardImpl impl_;
```
I think `VirtualGuardImpl` might be a more idiomatic way than `DeviceGuardImplInterface`, so I went for it.

lw commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -762,30 +732,62 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
-    std::vector<bool> isDeviceUsed(deviceCount, false);
-    for (const c10::Device& device : devices) {
+  // We need devices to be sorted in order to use ensureIsSubsetOfDevices.
+  static std::vector<c10::Device> sortAndDeduplicateDevices(
+      const c10::impl::VirtualGuardImpl& impl,
+      std::vector<c10::Device> devices) {
```
I might have ended up overengineering this function. I was just bothered that we were doing two heap allocations and throwing away the original vector when, in practice, we could often just use it as is.

I also wish we could just do `std::sort(...)` but we need to remove duplicates and that's tricky.

mrshenli replied:
This looks OK to me. Curious, will 1) sort devices + 2) two pointers/indices be a little shorter to implement?

lw replied:
That's a good idea!

mrshenli reviewed (approved):
Stamp, assuming that the added complexity for the CPU path does not lead to noticeable perf regression.

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -605,36 +584,32 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
-      return callback;
-    }
-
     return [this, callback=std::move(callback)]() {
```
NVM, I saw the callback wrapper is removed in the next PR.

lw committed:
Update on "Use Devices instead of DeviceIndexes in Future"


Even though we merged CUDAFuture into ivalue::Future, the resulting methods still had basically two distinct codepaths (i.e., an "early exit" if `impl_ == nullptr` for CPU, and then some code for CUDA). This works but it risks creating divergence and inconsistencies when the same class is used in those two modes. Ideally we should have the same codepath, and have the stream operations be no-ops for CPU. Luckily, this is exactly what happens when using a CPU DeviceGuardImplInterface!

Hence here I do that, and for convenience I also use c10::Devices instead of c10::DeviceIndexes (like we did in https://github.com/pytorch/pytorch/pull/57294 for RPC).

Differential Revision: [D28100525](https://our.internmc.facebook.com/intern/diff/D28100525/)

[ghstack-poisoned]

lw commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -561,37 +548,29 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
       const at::IValue& value,
       c10::optional<std::vector<std::reference_wrapper<const at::DataPtr>>>
           dataPtrs) {
-    if (impl_ == nullptr) {
```
Yes indeed. I admit I didn't benchmark as I assumed that ifs and empty loops would be negligible. Do you know some way I could measure this to make sure?

mrshenli replied:
I agree that most likely it won't lead to any noticeable differences in user land. So please feel free to skip measurements. The easiest way to measure would be creating a bunch of `torch.futures.Future` and mark them as completed. Though there are layers on top of the raw Future impl, but I think it should still be fine since this is how users gonna use it.

lw commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -318,16 +318,8 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
-    // an error. Uninitialized currentDevice_ could lead to crash when used
-    // in CUDAGuard.
-    if (impl_ != nullptr) {
-      currentDevice_ = impl_->getDevice().index();
```
According to the comment the reason for this was to avoid setting the current device to a "wrong" device. After this change, since we use an OptionalDeviceGuard, in such a case we would just not set the current device at all. This does indeed mean that the current device remains set to whatever it was before, hence it is indeed unspecified. Do you think this will be an issue?

lw committed:
Update on "Use Devices instead of DeviceIndexes in Future"


Even though we merged CUDAFuture into ivalue::Future, the resulting methods still had basically two distinct codepaths (i.e., an "early exit" if `impl_ == nullptr` for CPU, and then some code for CUDA). This works but it risks creating divergence and inconsistencies when the same class is used in those two modes. Ideally we should have the same codepath, and have the stream operations be no-ops for CPU. Luckily, this is exactly what happens when using a CPU DeviceGuardImplInterface!

Hence here I do that, and for convenience I also use c10::Devices instead of c10::DeviceIndexes (like we did in https://github.com/pytorch/pytorch/pull/57294 for RPC).

Differential Revision: [D28100525](https://our.internmc.facebook.com/intern/diff/D28100525/)

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in Future"


Even though we merged CUDAFuture into ivalue::Future, the resulting methods still had basically two distinct codepaths (i.e., an "early exit" if `impl_ == nullptr` for CPU, and then some code for CUDA). This works but it risks creating divergence and inconsistencies when the same class is used in those two modes. Ideally we should have the same codepath, and have the stream operations be no-ops for CPU. Luckily, this is exactly what happens when using a CPU DeviceGuardImplInterface!

Hence here I do that, and for convenience I also use c10::Devices instead of c10::DeviceIndexes (like we did in https://github.com/pytorch/pytorch/pull/57294 for RPC).

Differential Revision: [D28100525](https://our.internmc.facebook.com/intern/diff/D28100525/)

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in Future"


Even though we merged CUDAFuture into ivalue::Future, the resulting methods still had basically two distinct codepaths (i.e., an "early exit" if `impl_ == nullptr` for CPU, and then some code for CUDA). This works but it risks creating divergence and inconsistencies when the same class is used in those two modes. Ideally we should have the same codepath, and have the stream operations be no-ops for CPU. Luckily, this is exactly what happens when using a CPU DeviceGuardImplInterface!

Hence here I do that, and for convenience I also use c10::Devices instead of c10::DeviceIndexes (like we did in https://github.com/pytorch/pytorch/pull/57294 for RPC).

Differential Revision: [D28100525](https://our.internmc.facebook.com/intern/diff/D28100525/)

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in Future"


Even though we merged CUDAFuture into ivalue::Future, the resulting methods still had basically two distinct codepaths (i.e., an "early exit" if `impl_ == nullptr` for CPU, and then some code for CUDA). This works but it risks creating divergence and inconsistencies when the same class is used in those two modes. Ideally we should have the same codepath, and have the stream operations be no-ops for CPU. Luckily, this is exactly what happens when using a CPU DeviceGuardImplInterface!

Hence here I do that, and for convenience I also use c10::Devices instead of c10::DeviceIndexes (like we did in https://github.com/pytorch/pytorch/pull/57294 for RPC).

Differential Revision: [D28100525](https://our.internmc.facebook.com/intern/diff/D28100525/)

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in Future"


Even though we merged CUDAFuture into ivalue::Future, the resulting methods still had basically two distinct codepaths (i.e., an "early exit" if `impl_ == nullptr` for CPU, and then some code for CUDA). This works but it risks creating divergence and inconsistencies when the same class is used in those two modes. Ideally we should have the same codepath, and have the stream operations be no-ops for CPU. Luckily, this is exactly what happens when using a CPU DeviceGuardImplInterface!

Hence here I do that, and for convenience I also use c10::Devices instead of c10::DeviceIndexes (like we did in https://github.com/pytorch/pytorch/pull/57294 for RPC).

Differential Revision: [D28100525](https://our.internmc.facebook.com/intern/diff/D28100525/)

[ghstack-poisoned]
